lib/axon/compiler.ex

```diff
@@ -115,7 +134,17 @@ defmodule Axon.Compiler do
 
     init_cache = Map.new(cache, fn {_, {int_id, funs}} -> {int_id, funs} end)
 
-    init_fun = fn template, init_params ->
+    init_fun = fn template, init_state ->
+      # TODO: Check if init_state and output state match
+      {init_params, meta} =
+        case init_state do
+          %Axon.ModelState{data: init_params} = model_state ->
+            {init_params, Map.take(model_state, [:parameters, :frozen_parameters, :state])}
```

Comment:
If feels a bit backwards to tear the ModelState apart and then set it up again. What if you return `model_state` here and an empty model state in the branch below, and then always update it directly instead?

If you really want to break it apart, then I would break it apart into four distinct parameters:

```suggestion
            {init_params, model_state.parameters, model_state.frozen_parameters, model_state.state}
```